goserver/session/session_test.go

```diff
@@ -0,0 +1,70 @@
+package session
+
+import (
+	"net/http/httptest"
+	"reflect"
+	"strings"
+	"testing"
+)
+
+func TestNew(t *testing.T) {
+	sm := NewStoreManager(NewMemoryStore(), func() string { return "test-id" })
+	rr := httptest.NewRecorder()
+	r := httptest.NewRequest("method", "https://test.com", nil)
+	sess, err := New(sm, rr, r)
+	if err != nil {
+		t.Errorf("cannot create new session, got err: %v", err)
+	}
+
+	// Make sure session is created in manager.
+	sess, err = sm.Find(sess.ID)
+	if err != nil {
+		t.Fatalf("cannot find session in session manager, got err: %v", err)
+	}
+	if sess == nil {
+		t.Fatalf("cannot find session in session manager, got nil")
+	}
+
+	// Make sure cookie containing encoded session id is set in response.
+	cookies := rr.HeaderMap["Set-Cookie"]
+	if len(cookies) < 1 {
+		t.Fatal("cannot find cookie response")
+	}
+	esid := encodeSessionID(sess.ID)
+	if !strings.Contains(cookies[0], cookieName) {
+		t.Fatalf("cookie %s set in response", cookieName)
+	}
+	if !strings.Contains(cookies[0], esid) {
+		t.Fatal("no session id set in response cookie")
+	}
+
+	// Make sure cookie containing encoded session id is set in request.
```

Comment:
Related to Biao's question: why do we care that it's set in the request?

Reply:
Quote from my reply in Biao's question

"This is a standard way how session cookie should be handled (e.g., http://expressjs.com/en/resources/middleware/cookie-session.html)- the cookie should be set in the current request as well so that things like middlewares can be introduced."